test/e2e/storage/testsuites/topology.go

```diff
@@ -0,0 +1,320 @@
+		ok := false
+		dDriver, ok = driver.(DynamicPVTestDriver)
+		if !ok {
+			framework.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+		}
+
+		if !dInfo.Capabilities[CapTopology] {
+			framework.Skipf("Driver %q does not support topology - skipping", dInfo.Name)
+		}
+
+	})
+
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("topology")
+
+	init := func() {
+		l = topologyTest{}
+
+		// Now do the more expensive test initialization.
+		l.config, l.testCleanup = driver.PrepareTest(f)
+
+		l.resource = genericVolumeTestResource{
+			driver:  driver,
+			config:  l.config,
+			pattern: pattern,
+		}
+
+		// After driver is installed, check driver topologies on nodes
+		cs = f.ClientSet
+		keys := dInfo.TopologyKeys
+		if len(keys) == 0 {
+			framework.Skipf("Driver didn't provide topology keys -- skipping")
+		}
+		l.topologies, err = t.getCurrentTopologies(cs, keys, 2)
+		framework.ExpectNoError(err, "failed to get current driver topologies")
+		if len(l.topologies) < 2 {
+			framework.Skipf("Not enough topologies in cluster -- skipping")
+		}
+
+		l.resource.sc = dDriver.GetDynamicProvisionStorageClass(l.config, pattern.FsType)
+		framework.ExpectNotEqual(l.resource.sc, nil, "driver failed to provide a StorageClass")
+
+		l.resource.sc.VolumeBindingMode = &pattern.BindingMode
+		// Force all PVs to be provisioned in 1 topology
+		t.setAllowedTopologies(l.topologies[0], l.resource.sc)
+
+		claimSize := dDriver.GetClaimSize()
+		l.resource.pvc = framework.MakePersistentVolumeClaim(framework.PersistentVolumeClaimConfig{
+			ClaimSize:        claimSize,
+			StorageClassName: &(l.resource.sc.Name),
+		}, l.config.Framework.Namespace.Name)
+
+		l.intreeOps, l.migratedOps = getMigrationVolumeOpCounts(f.ClientSet, dInfo.InTreePluginName)
+	}
+
+	cleanup := func() {
+		t.cleanupResources(cs, &l)
+		if l.testCleanup != nil {
+			l.testCleanup()
+		}
+
+		validateMigrationVolumeOpCounts(f.ClientSet, dInfo.InTreePluginName, l.intreeOps, l.migratedOps)
+	}
+
+	ginkgo.It("should provision a volume and schedule a pod with AllowedTopologies", func() {
+		init()
+		defer func() {
+			cleanup()
+		}()
+
+		t.createResources(cs, &l, nil)
+
+		err = e2epod.WaitForPodRunningInNamespace(cs, l.pod)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Verifying pod scheduled to correct node")
+		pod, err := cs.CoreV1().Pods(l.pod.Namespace).Get(l.pod.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		node, err := cs.CoreV1().Nodes().Get(pod.Spec.NodeName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		t.verifyNodeTopology(node, l.topologies[0])
+	})
+
+	ginkgo.It("should fail to schedule a pod with conflicting AllowedTopologies [Slow]", func() {
```

Comment:
What about tests for delayed binding without `AllowedTopologies`? Are those covered somewhere else?